### PR TITLE
feat(rpc module): `stream API` for SubscriptionSink

### LINKED ISSUE #627: [rpc module]: SubscriptionSink more user-friendly API for sending items from stream.
We have noticed the API for `SubscriptionSink` is a bit messy to use if you want to await for stream to be sent via the `SubscriptionSink` if the connection has already been closed it might leak until the stream/fut "resolves" because the stream itself doesn't know anything of the connection state.

For example given that we have the following impl:

```rust
impl MyServerTrait for () 
{
	fn subscribe_to_something(&self, mut sink: SubscriptionSink) -> RpcResult<()> {
                // assume this stream lives forever and produces items "rarely"...
		let stream = create_stream();

		let fut = async move {
			stream
				.take_while(|sc| future::ready(sink.send(sc).map_or(false, |_| true)))
				.for_each(|_| future::ready(()))
				.await
		};

	       tokio::spawn(fut);
	}
}
```

If connection has already this case the `tokio task` will live until `sink.send` fails.
To be on the safe-side you would need to something like:

```rust
impl MyServerTrait for () {
    fn subscribe_to_something(&self, mut sink: SubscriptionSink) -> RpcResult<()> {
        // assume this stream lives forever and produces items "rarely"...
        let stream = create_stream();

        let fut = async move {
            loop {
                let timeout = tokio::time::sleep(std::time::Duration::from_secs(60));
                tokio::pin!(timeout);

                tokio::select! {
                    Some(item) = stream.next() => {
                        if let Err(e) = sink.send(&item) {
                          break;
                        }
                    },
                    _ = &mut timeout => {
                        if sink.is_closed() {
                            break;
                        }
                    }
                    else => break,
                };
            }
        };

        tokio::spawn(fut);
    }
}
```

### PR DESCRIPTION
**Pull request by niklasad1**
Fixes https://github.com/paritytech/jsonrpsee/issues/627

This PR introduces a new API for the `SubscriptionSink` to add a stream to send items over the subscription when "new items" get produced, if the connection gets closed it will terminate the task because it has additional channel linked to actual connection.

### CODE REVIEW

File: core/src/server/rpc_module.rs [hunk 1]
```diff
@@ -51,16 +51,35 @@ use serde::{de::DeserializeOwned, Serialize};
...
-	dyn Send + Sync + Fn(Id<'a>, Params<'a>, MethodSink, Option<ResourceGuard>, &dyn IdProvider) -> BoxFuture<'a, bool>,
+	dyn Send + Sync + Fn(Id<'a>, Params<'a>, MethodSink, ConnectionId, Option<ResourceGuard>) -> BoxFuture<'a, bool>,
 >;
 /// Connection ID, used for stateful protocol such as WebSockets.
 /// For stateless protocols such as http it's unused, so feel free to set it some hardcoded value.
 pub type ConnectionId = usize;
 /// Raw RPC response.
 pub type RawRpcResponse = (String, mpsc::UnboundedReceiver<String>, mpsc::UnboundedSender<String>);
+/// Connection state for stateful protocols such as WebSocket
+/// This is used to keep track whether the connection has been closed.
```
Comment (niklasad1) [inline core/src/server/rpc_module.rs:65]:
```suggestion
```

File: core/src/server/rpc_module.rs [hunk 2]
```diff
@@ -51,16 +51,34 @@ use serde::{de::DeserializeOwned, Serialize};
...
 pub type RawRpcResponse = (String, mpsc::UnboundedReceiver<String>, mpsc::UnboundedSender<String>);
+/// Connection state for stateful protocols such as WebSocket
+pub type MaybeConnState<'a> = Option<ConnState<'a>>;
+
+/// Data for stateful connections.
+pub struct ConnState<'a> {
+	/// Connection ID
+	pub conn_id: ConnectionId,
+	/// Channel to know whether the connection is closed or not.
+	pub close: async_channel::Receiver<()>,
```
Comment (niklasad1) [inline core/src/server/rpc_module.rs:72]: NOTE: [this panics if count > usize::MAX / 2](https://github.com/smol-rs/async-channel/blob/master/src/lib.rs#L692-#L696)

but if we reach that we likely have other problems such as OOM :)

File: tests/tests/integration_tests.rs [hunk 3]
```diff
@@ -359,6 +359,44 @@ async fn ws_server_should_stop_subscription_after_client_drop() {
 	assert!(matches!(close_err.close_reason(), &SubscriptionClosedReason::ConnectionReset));
 }
 
+#[tokio::test]
+async fn ws_server_cancel_stream_after_reset_conn() {
```
Comment (dvdplm) [inline tests/tests/integration_tests.rs:363]:
```suggestion
async fn ws_server_cancels_stream_after_reset_conn() {
```

File: tests/tests/integration_tests.rs [hunk 4]
```diff
@@ -359,6 +359,44 @@ async fn ws_server_should_stop_subscription_after_client_drop() {
...
+	let client = WsClientBuilder::default().build(&server_url).await.unwrap();
+	let _sub1: Subscription<usize> =
+		client.subscribe("subscribe_never_produce", None, "unsubscribe_never_produce").await.unwrap();
+	let _sub2: Subscription<usize> =
+		client.subscribe("subscribe_never_produce", None, "unsubscribe_never_produce").await.unwrap();
+
+	// terminate connection.
+	drop(client);
+	assert_eq!(Some(()), rx.next().await, "subscription stream should be terminated after the client was dropped");
+	assert_eq!(Some(()), rx.next().await, "subscription stream should be terminated after the client was dropped");
```
Comment (dvdplm) [inline tests/tests/integration_tests.rs:397]: I felt like I understood the test until I got here. I thought `rx` would produce `None` after the client was dropped and so it'd be `assert!(rx.next().await.is_none())`. What am I missing? :/

**Review by dvdplm — COMMENTED** [review]
This looks pretty good to me, what is it you want to do more here?

File: tests/tests/integration_tests.rs (hunk 4 above)
Comment (niklasad1) [reply tests/tests/integration_tests.rs:397]: haha, maybe it's more clear as you described it but the test actually sends a message on the channel when the subscription terminated.

**Comment by niklasad1** [comment]
> This looks pretty good to me, what is it you want to do more here?

It should be complete but the grumbles is maybe the API (naming) and cloning a bunch `async_channel::Receiver`...

File: core/src/server/rpc_module.rs [hunk 5]
```diff
@@ -426,8 +435,11 @@ impl Methods {
 	async fn inner_call(&self, req: Request<'_>) -> RawRpcResponse {
 		let (tx, mut rx) = mpsc::unbounded();
 		let sink = MethodSink::new(tx.clone());
+		let (_tx, rx_2) = async_channel::unbounded();
 
-		if let MethodResult::Async(fut) = self.execute(&sink, req, 0, &RandomIntegerIdProvider) {
+		let conn_state = Some(ConnState { conn_id: 0, close: rx_2, id_provider: &RandomIntegerIdProvider });
+
+		if let MethodResult::Async(fut) = self.execute(&sink, conn_state, req) {
```
Comment (niklasad1) [inline core/src/server/rpc_module.rs:442]: TODO: fix this to test the subscription closed properly which now works :)

File: core/src/server/rpc_module.rs [hunk 6]
```diff
@@ -786,9 +808,30 @@ impl SubscriptionSink {
...
+				// Stream or connection was dropped => close stream.
+				else => break,
+			}
+		}
+	}
+
 	/// Returns whether this channel is closed without needing a context.
 	pub fn is_closed(&self) -> bool {
-		self.inner.is_closed()
+		self.inner.is_closed() || self.close.is_closed()
```
Comment (niklasad1) [inline core/src/server/rpc_module.rs:834]: this is probably needless both should be dropped at the same time...

File: core/src/server/rpc_module.rs [hunk 7]
```diff
@@ -671,12 +683,13 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 			let subscribers = subscribers.clone();
 			self.methods.mut_callbacks().insert(
 				subscribe_method_name,
-				MethodCallback::new_sync(Arc::new(move |id, params, method_sink, conn_id, id_provider| {
+				MethodCallback::new_sync(Arc::new(move |id, params, method_sink, conn| {
 					let (conn_tx, conn_rx) = oneshot::channel::<()>();
+					let c = conn.expect("conn must be Some; this is bug");
```
Comment (maciejhirsz) [inline core/src/server/rpc_module.rs:688]: If it bust be `Some`, why don't we restrict it on parameter level?

File: core/src/server/rpc_module.rs [hunk 8]
```diff
@@ -157,24 +173,23 @@ impl MethodCallback {
...
 
 		let result = match &self.callback {
 			MethodKind::Sync(callback) => {
 				tracing::trace!(
-					"[MethodCallback::execute] Executing sync callback, params={:?}, req.id={:?}, conn_id={:?}",
+					"[MethodCallback::execute] Executing sync callback, params={:?}, req.id={:?}, conn_state={:?}",
 					params,
 					id,
-					conn_id
+					conn_state
```
Comment (maciejhirsz) [inline core/src/server/rpc_module.rs:189]: I'd rather unwrap `conn_state` here than later (as per other comment), but it's still not clear to me why it's mandatory here but optional for async calls.

Comment (niklasad1) [reply core/src/server/rpc_module.rs:189]: I moved `conn_id, id_provider, close` to `ConnState` before we required to pass `conn_id` explictly here which is unused.

Thus, for HTTP (with connection state) you would just pass `execute(&sink, req, None, resources)` however I get your point that this is error prone and you can't really `unwrap` here but `subscriptions` must be initialized with `Some(conn_state)` if that makes sense.

What do say about `execute_with_state` and `execute`? However I guess we would need additional callback types then.... :grimacing:

File: core/src/server/rpc_module.rs (hunk 7 above)
Comment (niklasad1) [reply core/src/server/rpc_module.rs:688]: it's way better but I think it requires additional callback types right? see my comment below.

thoughts?

Comment (niklasad1) [reply core/src/server/rpc_module.rs:688]: Ok, I had a look at the code to solve this properly at `parameter level` one need to lookup at the actual callback to avoid to pass down a bunch of unused parameters.

```rust
  // server code
  let result = match RpcModule::as_callback(&req.method) {
     None => {
        send_error(sink, msg),
        continue
     }
     Some(MethodKind::Sync(cb)) => (cb)(id, params, sink, conn_id),
     Some(MethodKind::Async(cb)) => (cb)(id, params, sink, resources).await,
     // should be used for subscriptions...
     // servers that don't support subscriptions should throw an error here...
     Some(MethodKind::Subscription) => (cb)(id, params, sink, resources, conn_state),
  };

  // modify register 
  pub fn register_subscription<F>(
     &mut self,
     subscribe_method_name: &'static str,
     notif_method_name: &'static str,
     unsubscribe_method_name: &'static str,
     callback: F,
 ) -> Result<(), Error> {
   .... 
   ....
   self.methods.mut_callbacks().insert(
      subscribe_method_name,
      MethodCallback::new_subscription(Arc::new(move |id, params, method_sink, conn| {
        ... 
     }
  );
 }


```

Comment (niklasad1) [reply core/src/server/rpc_module.rs:688]: one benefit is not to have to clone this "introduced channel connection_state" for every method instead only for subscriptions where it's actually used... not sure I liked how the old code abstracted this away

Comment (maciejhirsz) [reply core/src/server/rpc_module.rs]: I already thought about splitting async calls from subscriptions when doing my refactoring earlier, and for regular method calls returning values instead of passing in the sink as a parameter, I reckon that would make things much more readable and straight forward, and potentially make the final binary smaller. So if you want to go that route and add another enum variant I think that's cool, and I can do a PR that switches method calls to have a return value later on :).

Comment (niklasad1) [reply core/src/server/rpc_module.rs]: ah, I already added the enum variant in this PR but just a hacky draft to check that it works and to show you and David what I had in mind :)

Comment (niklasad1) [reply core/src/server/rpc_module.rs]: sounds great to get rid of the sink for the synchronous calls.

File: core/src/server/rpc_module.rs [hunk 9]
```diff
@@ -437,11 +459,18 @@ impl Methods {
...
 
-		let conn_state = Some(ConnState { conn_id: 0, close: close_rx, id_provider: &RandomIntegerIdProvider });
+		let id = req.id.clone();
+		let params = Params::new(req.params.map(|params| params.get()));
 
-		if let MethodResult::Async(fut) = self.execute(&sink, conn_state, req) {
-			fut.await;
-		}
+		let _result = match self.as_callback(&req.method).map(|c| &c.callback) {
+			None => todo!(),
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply core/src/server/rpc_module.rs:377]: exactly, forgot this good catch

File: ws-server/src/server.rs [hunk 10]
```diff
@@ -301,21 +306,23 @@ async fn background_task(
...
 		}
-		// terminate connection.
+
+		// Terminate connection and send close message.
 		let _ = sender.close().await;
-		// NOTE(niklasad1): when the receiver is dropped no further requests or subscriptions
-		// will be possible.
+
+		// Force `conn_tx` to this async block and close it down
+		// when the connection closes to be on safe side.
```
Comment (niklasad1) [inline ws-server/src/server.rs:324]: //cc @dvdplm this is what I meant by closing down the channel.

once this tasks terminates we drop `conn_tx` before I dropped it close/drop shouldn't matter just that it's moved into this block.

File: core/src/server/rpc_module.rs [hunk 11]
```diff
@@ -786,9 +746,40 @@ impl SubscriptionSink {
 		self.inner_send(msg).map_err(Into::into)
 	}
 
+	/// Consume the sink by passing a stream to be sent via the sink.
+	pub async fn add_stream<S, T>(mut self, mut stream: S)
+	where
+		S: Stream<Item = T> + Unpin,
+		T: Serialize,
+	{
+		let mut close_stream = self.close.clone();
```
Comment (niklasad1) [inline core/src/server/rpc_module.rs:755]: this is hack; to workaround borrowing issues when we call `self.send` below.

we may refactor this to split this up to several types or something....

File: core/src/server/rpc_module.rs [hunk 12]
```diff
@@ -786,9 +746,40 @@ impl SubscriptionSink {
 		self.inner_send(msg).map_err(Into::into)
 	}
 
+	/// Consume the sink by passing a stream to be sent via the sink.
+	pub async fn add_stream<S, T>(mut self, mut stream: S)
```
Comment (dvdplm) [inline core/src/server/rpc_module.rs:750]: What do you think about calling this `into_stream`? I think "add" implies there could be more than one and that it doesn't quite relay the information about the important changes that this call makes to the sink.

File: core/src/server/rpc_module.rs (hunk 11 above)
Comment (dvdplm) [reply core/src/server/rpc_module.rs:755]: I think we need plenty of comments in this method, it requires some pretty deep understanding of the architecture and I think we should capture it here.

File: ws-server/src/server.rs (hunk 10 above)
Comment (dvdplm) [reply ws-server/src/server.rs:324]: ok, ty

**Review by dvdplm — APPROVED** [review]

File: core/src/server/rpc_module.rs (hunk 12 above)
Comment (dvdplm) [reply core/src/server/rpc_module.rs:750]: Fair; I'd have liked `as_stream` but `as_` is "taken" with different semantics so can't do that.

Of your suggestions I like `from_stream` the best.

File: tests/tests/integration_tests.rs [hunk 13]
```diff
@@ -359,6 +359,44 @@ async fn ws_server_should_stop_subscription_after_client_drop() {
...
+	let mut module = RpcModule::new(tx);
+
+	module
+		.register_subscription("subscribe_never_produce", "n", "unsubscribe_never_produce", |_, sink, mut tx| {
+			// create stream that doesn't produce items.
+			let stream = futures::stream::empty::<usize>();
+			tokio::spawn(async move {
+				sink.add_stream(stream).await;
+				let send_back = Arc::make_mut(&mut tx);
+				send_back.feed(()).await.unwrap();
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply tests/tests/integration_tests.rs:380]: No, it's not the `SubscriptionSink` that is using `feed()`, it's the `mpsc::channel` to send messages back for testing purposes.

`feed` is another method on the `SinkExt trait` that doesn't require flushing as `send` does, https://docs.rs/futures/0.3.19/futures/sink/trait.SinkExt.html#method.feed. 

TLDR;  it's doesn't matter....

File: core/src/server/rpc_module.rs (hunk 12 above)
Comment (jsdw) [reply core/src/server/rpc_module.rs:750]: So, this method consumes a stream, feeding the items into the subscription?

I guess I'd go with something like `consume_stream` or `read_from_stream`. `into_`, `as_`, and `from_` all sortof feel like I should expect some result back from this call to me!

Comment (niklasad1) [reply core/src/server/rpc_module.rs:750]: I guess we could return a type that impls `Sink/SinkExt` instead here to make it more readable and flexible i.e, to deal with errors and so on.

Comment (dvdplm) [reply core/src/server/rpc_module.rs]: `streamify()`? 
I think `consume_stream` is so-so. Yes, we do consume it, but that's not really the point. Rather we're "hooking up" the stream to the sink and leave it there for the duration of the subscription.
`with_stream`?

File: core/src/server/rpc_module.rs [hunk 14]
```diff
@@ -786,9 +746,46 @@ impl SubscriptionSink {
...
+					match self.send(&i) {
+						Ok(_) => (),
+						Err(Error::SubscriptionClosed(_)) => break Ok(()),
+						Err(err) => break Err(err),
+					};
+					close = c;
+					item = stream.next();
+				}
+				// No messages should be sent over this channel
+				// if that's occur just continue.
```
Comment (niklasad1) [inline core/src/server/rpc_module.rs:775]:
```suggestion
				// if that occurs just continue.
```

File: core/src/server/rpc_module.rs (hunk 12 above)
Comment (jsdw) [reply core/src/server/rpc_module.rs]: `pipe`, maybe? we're piping a stream into the subscription

Comment (dvdplm) [reply core/src/server/rpc_module.rs]: `sink.pipe_from_stream`? I quite like `pipe`!

File: core/src/server/rpc_module.rs [hunk 15]
```diff
@@ -786,9 +746,71 @@ impl SubscriptionSink {
...
+				Either::Left((Some(i), c)) => {
+					match self.send(&i) {
+						Ok(_) => (),
+						Err(Error::SubscriptionClosed(close_reason)) => {
+							self.close(&close_reason);
+							break Ok(());
+						}
+						Err(err) => {
+							tracing::warn!("Subscription got error: {:?} terminating task", err);
+							break Err(err);
```
Comment (niklasad1) [inline core/src/server/rpc_module.rs:788]: this is most likely a bug when serializing T fails, should we have a flag for this when the exit, just log an continue, or something else?

The error is not really possible to handle but if you require each element to be processed in order, ignore might break expectations so I just terminated the "task" when an error is detected.

File: core/src/server/rpc_module.rs (hunk 12 above)
Comment (niklasad1) [reply core/src/server/rpc_module.rs]: I like `pipe_from_stream`, let's settle for that?

File: core/src/server/rpc_module.rs (hunk 15 above)
Comment (dvdplm) [reply core/src/server/rpc_module.rs:788]: I think you made the right choice. We could also panic I guess but let's terminate for now.

**Review by jsdw — APPROVED** [review]
Looks good to me overall, and I like the simplification in substrate!

**Review by maciejhirsz — APPROVED** [review]
Nice, good job separating subscriptions!
